### ISSUE TITLE
[R] MultiClass Training

### ISSUE THREAD

**Issue by melwinjose1991** [Microsoft/LightGBM#442]
Trying to train LGBM for a multiclass dataset. The below is the function
```
trainAndTestLGBM=function(nrounds=800, depth=8, leaves=200,  col_sample=0.5, 
                          row_sample=0.5, learn_rate=0.025, seed=1234, ltrain, lval, ltest){
  set.seed(seed)
  lxgb_params = list( objective = "multiclass", metric  = "multi_logloss",
                      learning_rate = learn_rate, 
                      num_leaves=leaves, max_depth = depth,  
                      bagging_fraction = row_sample, bagging_freq=1, bagging_seed = seed,
                      feature_fraction = col_sample, feature_fraction_seed = seed)
  
  lxgb_model = lgb.train(params=lxgb_params, data=ltrain, nrounds=nrounds, 
                valids=list(test = lval), 
                record=T, eval_freq = 50, num_class=3)
  
  all_predictions = predict(lxgb_model, data=ltest)
}
```

Getting the following Error:
```
Error in lgb.call(fun_name, ret = buf, ..., buf_len, act_len) : 
  api error: Number of classes should be specified and greater than 1 for multiclass training
Called from: lgb.call(fun_name, ret = buf, ..., buf_len, act_len)
```

How ltrain, lval and ltest are created
```
  y = as.integer(factor(data$y))   # has three classes 'A', 'B' and 'C'
  y = y - 1
  
  kfolds = 10
  folds = createFolds(y, k = kfolds, list = TRUE, returnTrain = FALSE)
  fold = as.numeric(unlist(folds[1]))
  
  x_train = data[-fold, x_cols] 
  x_val = data[fold, x_cols]
  
  y_train = y[-fold]
  y_val = y[fold]

  ltrain = lgb.Dataset(as.matrix(x_train), label=y_train)
  lval = lgb.Dataset(as.matrix(x_val), label=y_val)
  ltest = lgb.Dataset(as.matrix(test[, x_cols]))
```

**Comment by guolinke** [comment]
I think ```lgb.train``` doesn't have ```num_class``` parameter. 
you should set it in the ```lxgb_params ``` .

**Comment by melwinjose1991** [comment]
tried that as well, same error

**Comment by guolinke** [comment]
sorry. I thought it was in python package.
@Laurae2 any idea about this ?

**Comment by Laurae2** [comment]
@melwinjose1991 Without a reproducible example, will be difficult to troubleshoot this as I cannot reproduce it. Make sure you are using `lgb.unloader(wipe = TRUE)` to wipe LightGBM (and all its objects) from R before creating your matrices, it is a mandatory step if you messed up a training just before.

This demo works:

```r
library(lightgbm)

data(iris)

# We must convert factors to numeric
# They must be starting from number 0 to use multiclass
# For instance: 0, 1, 2, 3, 4, 5...
iris$Species <- as.integer(as.factor(iris$Species)) - 1

train <- as.matrix(iris[c(1:40, 51:90, 101:140), ])
test <- as.matrix(iris[c(41:50, 91:100, 141:150), ])
dtrain <- lgb.Dataset(data = train[, 1:4], label = train[, 5])
dtest <- lgb.Dataset.create.valid(dtrain, data = test[, 1:4], label = test[, 5])
valids <- list(test = dtest)

params <- list(objective = "multiclass", metric = "multi_logloss", num_class = 3)
model <- lgb.train(params,
                   dtrain,
                   100,
                   valids,
                   min_data = 1,
                   learning_rate = 1,
                   early_stopping_rounds = 10)

trainAndTestLGBM=function(nrounds=800, depth=8, leaves=200,  col_sample=0.5, 
                          row_sample=0.5, learn_rate=0.025, seed=1234, ltrain, lval, ltest){
  set.seed(seed)
  lxgb_params = list( objective = "multiclass", metric  = "multi_logloss",
                      learning_rate = learn_rate, 
                      num_leaves=leaves, max_depth = depth,  
                      bagging_fraction = row_sample, bagging_freq=1, bagging_seed = seed,
                      feature_fraction = col_sample, feature_fraction_seed = seed)
  
  lxgb_model = lgb.train(params=lxgb_params, data=ltrain, nrounds=nrounds, 
                         valids=list(test = lval), 
                         record=T, eval_freq = 50, num_class=3)
  
  all_predictions = predict(lxgb_model, data=ltest)
}

print(trainAndTestLGBM(nrounds = 3, ltrain = dtrain, lval = dtest, ltest = test))
```

Result (works even if using `dtest <- lgb.Dataset(data = test[, 1:4], label = test[, 5])`):

```r
> print(trainAndTestLGBM(nrounds = 3, ltrain = dtrain, lval = dtest, ltest = test))
[LightGBM] [Info] Total Bins 64
[LightGBM] [Info] Number of data: 120, number of used features: 4
[LightGBM] [Info] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] Trained a tree with leaves=2 and max_depth=1
[LightGBM] [Info] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] Trained a tree with leaves=2 and max_depth=1
[LightGBM] [Info] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] Trained a tree with leaves=2 and max_depth=1
[1]:	test's multi_logloss:1.08058 
[LightGBM] [Info] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] Trained a tree with leaves=2 and max_depth=1
[LightGBM] [Info] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] Trained a tree with leaves=2 and max_depth=1
[LightGBM] [Info] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] Trained a tree with leaves=2 and max_depth=1
[LightGBM] [Info] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] Trained a tree with leaves=2 and max_depth=1
[LightGBM] [Info] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] Trained a tree with leaves=2 and max_depth=1
[LightGBM] [Info] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] Trained a tree with leaves=2 and max_depth=1
[3]:	test's multi_logloss:1.03655 
 [1] 0.3655953 0.3162812 0.3181235 0.3655953 0.3162812 0.3181235 0.3655953 0.3162812 0.3181235 0.3376158 0.3347418 0.3276424 0.3655953 0.3162812
[15] 0.3181235 0.3655953 0.3162812 0.3181235 0.3655953 0.3162812 0.3181235 0.3655953 0.3162812 0.3181235 0.3655953 0.3162812 0.3181235 0.3655953
[29] 0.3162812 0.3181235 0.3312695 0.3472469 0.3214836 0.3196981 0.3473304 0.3329716 0.3312695 0.3472469 0.3214836 0.3312695 0.3472469 0.3214836
[43] 0.3312695 0.3472469 0.3214836 0.3312695 0.3472469 0.3214836 0.3312695 0.3472469 0.3214836 0.3233882 0.3513394 0.3252725 0.3312695 0.3472469
[57] 0.3214836 0.3312695 0.3472469 0.3214836 0.3095121 0.3362640 0.3542240 0.3095121 0.3362640 0.3542240 0.3172138 0.3325133 0.3502730 0.3095121
[71] 0.3362640 0.3542240 0.3095121 0.3362640 0.3542240 0.3095121 0.3362640 0.3542240 0.3095121 0.3362640 0.3542240 0.3095121 0.3362640 0.3542240
[85] 0.3095121 0.3362640 0.3542240 0.3095121 0.3362640 0.3542240
```

Even works when the number of classes do not match, like this:

```r
train <- as.matrix(iris[c(1:40, 51:90, 141:150), ])
```

Or this:

```r
test <- as.matrix(iris[c(41:50, 91:100), ])
```

**Comment by melwinjose1991** [comment]
@Laurae2 i just restart R studio and it worked. Maybe lgb.unloader() would also have worked. 
One last question, how to turn off just the [info] part of the verbose ?
